File: SpotifyAPI/SpotifyAPI.BLL/Services/Track/TrackService.cs
using AutoMapper;
using AutoMapper.QueryableExtensions;
using Microsoft.EntityFrameworkCore;
using SpotifyAPI.BLL.DTOs;
using SpotifyAPI.BLL.DTOs.Track;
using SpotifyAPI.DAL.Entities;
using SpotifyAPI.DAL.Repositories;
using SpotifyAPI.DAL.Repositories.Track;

namespace SpotifyAPI.BLL.Services.Track;

public class TrackService : ITrackService
{
    private readonly ITrackRepository _repository;

    public TrackService(ITrackRepository repository)
    {
        _repository = repository;
    }
    
    public async Task<ServiceResponse> GetPagedAsync(string userId, int page, int pageSize)
    {
        var tracksQuery = _repository
            .GetAll()
            .Select(t => new TrackDto
            {
                Id = t.Id,
                Name = t.Name,
                Duration = t.Duration,
                Path = t.Path,
                Image = t.Image,
                Artist = t.Album.Artist.Name,
                IsLiked = t.Likes.Any(l => l.UserId == userId)
            });
        
        var paged = await tracksQuery.ToPageAsync(new PageQuery
        {
            Page = page,
            PageSize = pageSize
        });
        
        if (paged.Total == 0)
            return new ServiceResponse("Tracks not found");
        
        return new ServiceResponse("Tracks loaded", true, paged);
    }

    public async Task<ServiceResponse> RecentlyPlayedAsync(string userId, int count)
    {
        var recentlyPlayed = await _repository
            .RecentlyPlayed(userId, count)
            .Select(p => new TrackDto
            {
                Id = p.Track!.Id,
                Name = p.Track!.Name,
                Duration = p.Track!.Duration,
                Path = p.Track!.Path,
                Image = p.Track!.Image,
                Artist = p.Track!.Album!.Artist!.Name,
                IsLiked = p.Track.Likes.Any(l => l.UserId == userId),
            }).ToListAsync();
        
        if (recentlyPlayed.Count == 0)
            return new ServiceResponse("Tracks not found");
        
        return new ServiceResponse("Tracks loaded", true, recentlyPlayed);
    }

    public async Task<ServiceResponse> CreateRecentlyPlayedAsync(string userId, string trackId)
    {
        if (string.IsNullOrWhiteSpace(userId) || string.IsNullOrWhiteSpace(trackId))
            return new ServiceResponse("Invalid userId or trackId");

        var ok = await _repository.UpsertAsync(userId, trackId);
        
        if (!ok) 
            return new ServiceResponse("Track not found or cannot create recently played");

        return new ServiceResponse("Recently played saved", true);
    }

    public async Task<ServiceResponse> GetByAlbumAsync(string userId, string albumId)
    {
        var tracks = await _repository
            .GetByAlbum(albumId)
            .Select(t => new TrackDto
            {
                Id = t.Id,
                Name = t.Name,
                Duration = t.Duration,
                Path = t.Path,
                Image = t.Image,
                Artist = t.Album.Artist.Name,
                IsLiked = t.Likes.Any(l => l.UserId == userId)
            })
            .ToListAsync();
        
        if (tracks.Count == 0)
            return new ServiceResponse("Tracks not found", false, tracks);
        
        return new ServiceResponse("Tracks loaded", true, tracks);
    }

    public async Task<ServiceResponse> GetLikedAsync(string userId, int page, int pageSize)
    {
        var tracksQuery = _repository
            .GetLiked(userId)
            .Select(t => new TrackDto
            {
                Id = t.Id,
                Name = t.Name,
                Duration = t.Duration,
                Path = t.Path,
                Image = t.Image,
                Artist = t.Album.Artist.Name,
                IsLiked = t.Likes.Any(l => l.UserId == userId)
            });
        
        var paged = await tracksQuery.ToPageAsync(new PageQuery
        {
            Page = page,
            PageSize = pageSize
        });
        
        if (paged.Total == 0)
            return new ServiceResponse("Tracks not found");
        
        return new ServiceResponse("Tracks loaded", true, paged);
    }

    public async Task<ServiceResponse> GetByQueryAsync(string userId, string query, int count)
    {
        var tracks = await _repository
            .GetByQuery(query, count)
            .Select(t => new TrackDto
            {
                Id = t.Id,
                Name = t.Name,
                Duration = t.Duration,
                Path = t.Path,
                Image = t.Image,
                Artist = t.Album.Artist.Name,
                IsLiked = t.Likes.Any(l => l.UserId == userId)
            })
            .ToListAsync();
        
        if (tracks.Count == 0)
            return new ServiceResponse("Tracks not found", false, tracks);
        
        return new ServiceResponse("Tracks loaded", true, tracks);
    }

    public async Task<ServiceResponse> GetByGenreAsync(string userId, string genre, int page, int pageSize)
    {
        var tracksQuery = _repository
            .GetByGenre(genre)
            .Select(t => new TrackDto
            {
                Id = t.Id,
                Name = t.Name,
                Duration = t.Duration,
                Path = t.Path,
                Image = t.Image,
                Artist = t.Album.Artist.Name,
                IsLiked = t.Likes.Any(l => l.UserId == userId)
            });
        
        var paged = await tracksQuery.ToPageAsync(new PageQuery
        {
            Page = page,
            PageSize = pageSize
        });
        
        if (paged.Total == 0)
            return new ServiceResponse("Tracks not found");
        
        return new ServiceResponse("Tracks loaded", true, paged);
    }
}

File: SpotifyReact/src/api/services/trackService.ts
import { apiFetch } from "../apiClient";
import { unwrapServiceResponse } from "../unwrapServiceResponse";

export async function getQuickPicks(page = 1, pageSize = 20) {
  const res = await apiFetch(`/api/track/paged?page=${page}&pageSize=${pageSize}`, {
    method: "GET",
  });

  const payload = await unwrapServiceResponse<any>(res);
  return payload?.items ?? [];
}

export async function getQuickPicksPaged(page = 1, pageSize = 20) {
  const res = await apiFetch(`/api/track/paged?page=${page}&pageSize=${pageSize}`, {
    method: "GET",
  });

  return await unwrapServiceResponse<any>(res);
}

export async function getRecentlyPlayed(userId: string) {
  const res = await apiFetch(
    `/api/track/recently-played?userId=${encodeURIComponent(userId)}`,
    { method: "GET" }
  );

  const payload = await unwrapServiceResponse<any>(res);
  return payload ?? [];
}

export async function getLiked(page = 1, pageSize = 10) {
  const res = await apiFetch(
    `/api/track/favourite?page=${page}&pageSize=${pageSize}`,
    { method: "GET" }
  );

  return await unwrapServiceResponse<any>(res);
}

export async function getTracksByQuery(query: string, count = 10) {
  const res = await apiFetch(
    `/api/track/query?query=${query}&count=${count}`,
    { method: "GET" }
  );

  const payload = await unwrapServiceResponse<any>(res);
  return payload ?? [];
}

export async function getTracksByAlbum(albumId: string) {
  const res = await apiFetch(
    `/api/track/album?albumId=${encodeURIComponent(albumId)}`,
    { method: "GET" }
  );

  const payload = await unwrapServiceResponse<any>(res);
  return payload ?? [];
}

export async function getTracksByGenre(genre: string, page = 1, pageSize = 10) {
  const res = await apiFetch(
    `/api/track/genre?genre=${genre}&page=${page}&pageSize=${pageSize}`,
    { method: "GET" }
  );

  const payload = await unwrapServiceResponse<any>(res);
  return payload.items ?? [];
}